lm_human_preference_details/train_reward_jax.py

```diff
@@ -0,0 +1,765 @@
+    gradient_accumulation_steps: int = 1
+    """gradient accumulation steps"""
+    local_micro_batch_size: tyro.conf.Suppress[int] = None
+    """per rank micro batch size"""
+    lr: float = 0.00005
+    """the learning rate"""
+    eps: float = 1e-5
+    """the epsilon for AdamW"""
+    rollout_batch_size: int = 512  # decrease this to e.g. 64 if OOM
+    """rollout batch size"""
+    world_size: tyro.conf.Suppress[int] = None
+    """the number of processes to use"""
+    batch_size: tyro.conf.Suppress[int] = None
+    """the batch size across all ranks"""
+    local_normalize_samples: int = 256
+    """Samples used to estimate reward mean and std"""
+    normalize_samples: tyro.conf.Suppress[int] = None
+    """Samples used to estimate reward mean and std across all ranks"""
+    debug_normalize: int = 0
+    """Samples used to check that normalization worked"""
+    normalize_before: bool = True
+    """Whether, before training, to normalize the rewards on the policy to the
+    scales on the training buffer. (For comparisons, just use mean 0, var 1.)"""
+    normalize_after: bool = True
+    """Whether, after training, to normalize the rewards on the ref policy to
+    mean 0, var 1 (so the KL coefficient always has the same meaning)."""
+    print_sample_output_freq: int = 10
+    """How often to print sample output"""
+    save_path: str = "models/"
+    """Where to save the model"""
+    use_tensorflow_adam: bool = False
+    """Whether to use tensorflow-style Adam optimizer instead of PyTorch's"""
+    task: TaskHParams = field(default_factory=TaskHParams)
+    labels: LabelHParams = field(default_factory=LabelHParams)
+
+
+OPENAI_PAD_TOKEN_ID = 50259
+
+
+@flax.struct.dataclass
+class RewardModelParams:
+    """Parameters for the reward model."""
+
+    backbone_params: flax.core.FrozenDict
+    head_params: flax.core.FrozenDict
+
+
+class RewardHead(nn.Module):
+    """Affine transform head for the reward model.
+
+    Attributes:
+      head_input_size: Size of the input to the head. The weights of the linear
+        layer are initialized from mean-zero Gaussians with std
+        1/sqrt(head_input_size + 1).
+
+    Example:
+      model = RewardHead(head_input_size=768)
+      variables = model.init(jax.random.PRNGKey(0),
+        jnp.ones((1, 6, 768)))
+    """
+
+    head_input_size: int
+
+    def setup(self):
+        self.reward_linear = nn.Dense(
+            1,
+            kernel_init=nn.initializers.normal(
+                stddev=1 / np.sqrt(self.head_input_size + 1)
+            ),
+            bias_init=nn.initializers.zeros_init(),
+        )
+        self.reward_gain = self.param("reward_gain", nn.initializers.ones_init(), ())
+        self.reward_bias = self.param("reward_bias", nn.initializers.zeros_init(), ())
+
+    def __call__(self, x):
+        assert x.shape[-1] == self.head_input_size
+        x = self.reward_linear(x)
+        x = x * self.reward_gain + self.reward_bias
+        return x
```

Comment:
My ideal format looks like the following, where we eliminate the need for `RewardModelParams`

```

class AutoModelForCausalLMWithRewardHead(nn.Module):
    model_name: str

    @nn.compact
    def __call__(self, x):
        backbone = transformers.FlaxAutoModelForCausalLM.from_pretrained(
            self.model_name
        )
        x = backbone(
            input_ids=x,
            # attention_mask=np.ones(x),
            # position_ids=xxx,
            return_dict=True,
            output_hidden_states=True,
        ).logits
        x = nn.Dense(
            1,
            kernel_init=nn.initializers.normal(
                stddev=1 / jnp.sqrt(backbone.config.hidden_size + 1)
            ),
            bias_init=nn.initializers.zeros_init(),
        )(x)
        reward_gain = self.param("reward_gain", nn.initializers.ones_init(), ())
        reward_bias = self.param("reward_bias", nn.initializers.zeros_init(), ())
        x = x * reward_gain + reward_bias
        return x
    
reward_model = AutoModelForCausalLMWithRewardHead("gpt2")
params = reward_model.init(
    jax.random.PRNGKey(0),
    jnp.array([
        [11, 339, 561],
    ]),
)
```

However, this only initialize the params for the reward model and the backbone params are not included. 


Maybe for now we can change it to

```

class RewardHead(nn.Module):
    head_input_size: int

    @nn.compact
    def __call__(self, x):
        assert x.shape[-1] == self.head_input_size
        x = nn.Dense(
            1,
            kernel_init=nn.initializers.normal(
                stddev=1 / np.sqrt(self.head_input_size + 1)
            ),
            bias_init=nn.initializers.zeros_init(),
        )(x)
        reward_gain = self.param("reward_gain", nn.initializers.ones_init(), ())
        reward_bias = self.param("reward_bias", nn.initializers.zeros_init(), ())
        x = x * reward_gain + reward_bias
        return x
```